### Regan Alt - Assignment 2 - 06/15/2021

In [1]:
import sqlite3 as sql
import pandas as pd

#### PART A: Query tickers with fetchall()

In [2]:
# PART A: Query tickers with fetchall()
# Write Python statements to query the tickers table in the dj.db database. Use the SQL statement: SELECT id, ticker, name FROM tickers
# place SQL results in a DataFrame
import sqlite3 as sql
import pandas as pd
connection = sql.Connection("dj.db")
cursor = connection.cursor()
cursor.execute("SELECT id, ticker, name FROM tickers")
rows = cursor.fetchall()
data = pd.DataFrame(rows) # simple Dataframe creation since it can take a matrix
data

0     1                                            2
0    1     M                                   Macy's Inc
1    2   AXP                     American Express Company
2    3  AAPL                                   Apple Inc.
3    4    BA                           The Boeing Company
4    5   CAT                             Caterpillar Inc.
5    6   CVX                          Chevron Corporation
6    7  CSCO                           Cisco Systems Inc.
7    8    KO                      Coca-Cola Company (The)
8    9   DIS                      The Walt Disney Company
9   10   DOW                                     Dow Inc.
10  11   XOM                      Exxon Mobil Corporation
11  12    GS               Goldman Sachs Group Inc. (The)
12  13    HD                        Home Depot Inc. (The)
13  14   IBM  International Business Machines Corporation
14  15  INTC                            Intel Corporation
15  16   JNJ                            Johnson & Johnson
16  17   JPM                        JP Morgan Chase & Co.
17  18   MCD                       McDonald's Corporation
18  19   MRK                   Merck & Company Inc. (new)
19  20  MSFT                        Microsoft Corporation
20  21   NKE                                    Nike Inc.
21  22   PFE                                  Pfizer Inc.
22  23    PG               Procter & Gamble Company (The)
23  24   TRV                 The Travelers Companies Inc.
24  25   UTX              United Technologies Corporation
25  26   UNH         UnitedHealth Group Incorporated (DE)
26  27    VZ                  Verizon Communications Inc.
27  28     V                                    Visa Inc.
28  29   WMT                                 Walmart Inc.
29  30   WBA                Walgreens Boots Alliance Inc.

In [3]:
# in the DataFrame constructed above, give meaningful column names
data.columns = ['id', 'ticker', 'name']
data

id ticker                                         name
0    1      M                                   Macy's Inc
1    2    AXP                     American Express Company
2    3   AAPL                                   Apple Inc.
3    4     BA                           The Boeing Company
4    5    CAT                             Caterpillar Inc.
5    6    CVX                          Chevron Corporation
6    7   CSCO                           Cisco Systems Inc.
7    8     KO                      Coca-Cola Company (The)
8    9    DIS                      The Walt Disney Company
9   10    DOW                                     Dow Inc.
10  11    XOM                      Exxon Mobil Corporation
11  12     GS               Goldman Sachs Group Inc. (The)
12  13     HD                        Home Depot Inc. (The)
13  14    IBM  International Business Machines Corporation
14  15   INTC                            Intel Corporation
15  16    JNJ                            Johnson & Johnson
16  17    JPM                        JP Morgan Chase & Co.
17  18    MCD                       McDonald's Corporation
18  19    MRK                   Merck & Company Inc. (new)
19  20   MSFT                        Microsoft Corporation
20  21    NKE                                    Nike Inc.
21  22    PFE                                  Pfizer Inc.
22  23     PG               Procter & Gamble Company (The)
23  24    TRV                 The Travelers Companies Inc.
24  25    UTX              United Technologies Corporation
25  26    UNH         UnitedHealth Group Incorporated (DE)
26  27     VZ                  Verizon Communications Inc.
27  28      V                                    Visa Inc.
28  29    WMT                                 Walmart Inc.
29  30    WBA                Walgreens Boots Alliance Inc.

In [4]:
# for DataFrame above, the id column is a good choice for DataFrame
data.set_index('id', inplace=True)
data

ticker                                         name
id                                                    
1       M                                   Macy's Inc
2     AXP                     American Express Company
3    AAPL                                   Apple Inc.
4      BA                           The Boeing Company
5     CAT                             Caterpillar Inc.
6     CVX                          Chevron Corporation
7    CSCO                           Cisco Systems Inc.
8      KO                      Coca-Cola Company (The)
9     DIS                      The Walt Disney Company
10    DOW                                     Dow Inc.
11    XOM                      Exxon Mobil Corporation
12     GS               Goldman Sachs Group Inc. (The)
13     HD                        Home Depot Inc. (The)
14    IBM  International Business Machines Corporation
15   INTC                            Intel Corporation
16    JNJ                            Johnson & Johnson
17    JPM                        JP Morgan Chase & Co.
18    MCD                       McDonald's Corporation
19    MRK                   Merck & Company Inc. (new)
20   MSFT                        Microsoft Corporation
21    NKE                                    Nike Inc.
22    PFE                                  Pfizer Inc.
23     PG               Procter & Gamble Company (The)
24    TRV                 The Travelers Companies Inc.
25    UTX              United Technologies Corporation
26    UNH         UnitedHealth Group Incorporated (DE)
27     VZ                  Verizon Communications Inc.
28      V                                    Visa Inc.
29    WMT                                 Walmart Inc.
30    WBA                Walgreens Boots Alliance Inc.

### Part B – Determine the date span of the data

Write Python statements to query the daily_measures table in the dj.db database, to determine the
earliest and latest date of the data in this table

In [5]:
# place SQL results in a DataFrame - method 2: pandas read_sql_query()
# use named parameter index_col
# Use SQL to join 2 tables; Python statements are similar to one-table queries
import sqlite3 as sql
connection = sql.Connection("dj.db")
cursor = connection.cursor()
query = """
 SELECT min(date) as first, max(date) as last FROM daily_measures
"""
cursor.execute(query)
rows = cursor.fetchall()
print(rows)

[('2017-01-03', '2017-12-29')]


### Answer: The earliest date in this data is January 3rd, 2017, and the last data is December 29th, 2017. 

## Part C – Query price history with for and Cursor

Write Python statements to query the daily_measures table in the dj.db database, and based on the
returned data, determine the date and closing price of the highest price for Apple stock during the data
available in the database.

In [6]:
# fetchone() used here with for
import sqlite3 as sql
connection = sql.Connection("dj.db")
cursor = connection.cursor()
cursor.execute(" SELECT date, closing FROM daily_measures WHERE ticker_id=3 ORDER BY date")
for row in cursor: # Cursor provides a shorthand way to iterate over the rows
   print(row)

('2017-01-03', 114.715378)
('2017-01-04', 114.586983)
('2017-01-05', 115.169696)
('2017-01-06', 116.453639)
('2017-01-09', 117.5203)
('2017-01-10', 117.638817)
('2017-01-11', 118.270913)
('2017-01-12', 117.777088)
('2017-01-13', 117.569682)
('2017-01-17', 118.517825)
('2017-01-18', 118.507948)
('2017-01-19', 118.300542)
('2017-01-20', 118.517825)
('2017-01-23', 118.596837)
('2017-01-24', 118.488195)
('2017-01-25', 120.374604)
('2017-01-26', 120.433863)
('2017-01-27', 120.443739)
('2017-01-30', 120.127692)
('2017-01-31', 119.85115)
('2017-02-01', 127.159749)
('2017-02-02', 126.942467)
('2017-02-03', 127.485673)
('2017-02-06', 128.680728)
('2017-02-07', 129.905412)
('2017-02-08', 130.409113)
('2017-02-09', 131.347379)
('2017-02-10', 131.049809)
('2017-02-13', 132.210332)
('2017-02-14', 133.926319)
('2017-02-15', 134.41235)
('2017-02-16', 134.248686)
('2017-02-17', 134.620649)
('2017-02-21', 135.592711)
('2017-02-22', 135.99939)
('2017-02-23', 135.424088)
('2017-02-24', 135.553035)
('2017

In [7]:
#Find the date of Apple's maximum closing price
import sqlite3 as sql
connection = sql.Connection("dj.db")
cursor = connection.cursor()
cursor.execute(" SELECT date, max(closing) FROM daily_measures WHERE ticker_id=3 ORDER BY date")
for row in cursor: # Cursor provides a shorthand way to iterate over the rows
   print(row)

('2017-12-18', 176.42)


### Answer: Apple's max closing price was 176.42 dollars, on December 18, 2017.

## Part D – Query price history to determine largest 1-day price percentage moves

Write Python statements to determine the date and price when Apple had the highest one-day daily
percent change (See Appendix B for definition). Use the same SQL statement as in part C, and
additional Python (and possibly pandas) statements to find the result.

In [8]:
# fetchone() used here with for
# previous SQL statement from part C
import sqlite3 as sql
connection = sql.Connection("dj.db")
cursor = connection.cursor()
cursor.execute(" SELECT date, closing FROM daily_measures WHERE ticker_id=3 ORDER BY date")
for row in cursor: # Cursor provides a shorthand way to iterate over the rows
   print(row)

('2017-01-03', 114.715378)
('2017-01-04', 114.586983)
('2017-01-05', 115.169696)
('2017-01-06', 116.453639)
('2017-01-09', 117.5203)
('2017-01-10', 117.638817)
('2017-01-11', 118.270913)
('2017-01-12', 117.777088)
('2017-01-13', 117.569682)
('2017-01-17', 118.517825)
('2017-01-18', 118.507948)
('2017-01-19', 118.300542)
('2017-01-20', 118.517825)
('2017-01-23', 118.596837)
('2017-01-24', 118.488195)
('2017-01-25', 120.374604)
('2017-01-26', 120.433863)
('2017-01-27', 120.443739)
('2017-01-30', 120.127692)
('2017-01-31', 119.85115)
('2017-02-01', 127.159749)
('2017-02-02', 126.942467)
('2017-02-03', 127.485673)
('2017-02-06', 128.680728)
('2017-02-07', 129.905412)
('2017-02-08', 130.409113)
('2017-02-09', 131.347379)
('2017-02-10', 131.049809)
('2017-02-13', 132.210332)
('2017-02-14', 133.926319)
('2017-02-15', 134.41235)
('2017-02-16', 134.248686)
('2017-02-17', 134.620649)
('2017-02-21', 135.592711)
('2017-02-22', 135.99939)
('2017-02-23', 135.424088)
('2017-02-24', 135.553035)
('2017

In [21]:
# Find Closing Price percent change by converting SQL query of Price History to a pandas dataframe
import pyodbc
import pandas as pd

df_adjprice = sql.Connection("dj.db")

SQL_Query = pd.read_sql_query(
'''select
date,
closing
from daily_measures WHERE ticker_id=3 ORDER BY date''', df_adjprice)

df = pd.DataFrame(SQL_Query, columns=['ticker','date','closing'])
print (df)

     ticker        date     closing
0       NaN  2017-01-03  114.715378
1       NaN  2017-01-04  114.586983
2       NaN  2017-01-05  115.169696
3       NaN  2017-01-06  116.453639
4       NaN  2017-01-09  117.520300
..      ...         ...         ...
244     NaN  2017-12-22  175.010000
245     NaN  2017-12-26  170.570000
246     NaN  2017-12-27  170.600000
247     NaN  2017-12-28  171.080000
248     NaN  2017-12-29  169.230000

[249 rows x 3 columns]


In [22]:
df["Pct Change"] = df["closing"].pct_change()
df

ticker        date     closing  Pct Change
0       NaN  2017-01-03  114.715378         NaN
1       NaN  2017-01-04  114.586983   -0.001119
2       NaN  2017-01-05  115.169696    0.005085
3       NaN  2017-01-06  116.453639    0.011148
4       NaN  2017-01-09  117.520300    0.009160
..      ...         ...         ...         ...
244     NaN  2017-12-22  175.010000    0.000000
245     NaN  2017-12-26  170.570000   -0.025370
246     NaN  2017-12-27  170.600000    0.000176
247     NaN  2017-12-28  171.080000    0.002814
248     NaN  2017-12-29  169.230000   -0.010814

[249 rows x 4 columns]

In [26]:
df.loc[df['Pct Change'].idxmax()]

ticker               NaN
date          2017-02-01
closing           127.16
Pct Change     0.0609806
Name: 20, dtype: object

### Answer: The largest 1 day closing price percentage change for Apple in 2017 took place on February 1st, 2017 with a percent change of 0.0609 or about a 6% increase from the previous day.